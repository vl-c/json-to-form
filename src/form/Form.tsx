import * as React from 'react';
import BaseInput, { BaseInputProps } from '../fields/BaseInput';
import TextArea, { TextAreaProps } from '../fields/TextArea';
import CheckboxInput, { CheckboxInputProps } from '../fields/CheckboxInput';
import RadioButtons, { RadioButtonsProps } from '../fields/RadioButtons';
import DateInput, { DateInputProps } from '../fields/DateInput';

export interface FormSchema {
  header?: string;
  buttons?: {
    id: string;
    type: 'button' | 'reset' | 'submit';
    value: string;
  }[];
  items?: (
    | BaseInputProps
    | (TextAreaProps & { type: string })
    | (CheckboxInputProps & { type: string })
    | (RadioButtonsProps & { type: string })
    | (DateInputProps & { type: string })
  )[];
}
interface FormProps {
  formSchema: FormSchema | null;
  submitHandler?: (res: { [key: string]: FormDataEntryValue }) => void;
}

const Form = ({ formSchema, submitHandler }: FormProps): JSX.Element => {
  const { header, items, buttons } = formSchema || {};

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    const formData = new FormData(e.target as HTMLFormElement);
    // just example
    const res: { [key: string]: FormDataEntryValue } = {};
    formData.forEach((value, key) => (res[key] = value));
    if (submitHandler) {
      submitHandler(res);
    } else {
      console.log(res);
    }
  };

  const renderButtons = ({
    id,
    type,
    value,
  }: {
    id: string;
    type: string;
    value: string;
  }) => <input key={id} type={type} value={value} />;

  const renderField = (item: {
    id: string;
    type: string;
  }): JSX.Element | null => {
    switch (item.type) {
      case 'text':
      case 'number':
        return <BaseInput key={item.id} {...(item as BaseInputProps)} />;
      case 'textarea':
        return <TextArea key={item.id} {...(item as TextAreaProps)} />;
      case 'checkbox':
        return (
          <CheckboxInput key={item.id} {...(item as CheckboxInputProps)} />
        );
      case 'radio':
        return (
          <RadioButtons
            key={item.id}
            {...((item as unknown) as RadioButtonsProps)}
          />
        );
      case 'date':
        return <DateInput key={item.id} {...(item as DateInputProps)} />;
      default:
        return null;
    }
  };

  return (
    <>
      {header && <h2>{header}</h2>}
      <form onSubmit={handleSubmit}>
        {items && items.map(renderField)}
        <div className="input-buttons">
          {buttons && buttons.map(renderButtons)}
        </div>
      </form>
    </>
  );
};
export default Form;
