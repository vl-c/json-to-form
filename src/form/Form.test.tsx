import * as React from 'react';

import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Form from './Form';

const formSchema = {
  items: [
    {
      label: 'count',
      type: 'number',
      id: 'first',
    },
    {
      label: 'DATE',
      type: 'date',
      id: 'datefield',
      value: '2021-04-08T23:35:58.282Z',
    },
    {
      label: 'name',
      value: 'hello',
      type: 'text',
      id: 'second',
    },
    {
      label: 'textarea',
      value: 'many text',
      type: 'textarea',
      id: 'third',
    },
    {
      label: 'checkbox',
      value: true,
      type: 'checkbox',
      id: 'cb',
    },
    {
      id: 'typeOfContact',
      type: 'radio',
      items: [
        { value: 'mail', label: 'mail' },
        { value: 'phone', label: 'phone' },
      ],
    },
  ],
};

function renderForm() {
  return shallow(<Form formSchema={formSchema} />);
}

it('Form shallow render', () => {
  const res = renderForm();
  expect(shallowToJson(res)).toMatchSnapshot();
});
