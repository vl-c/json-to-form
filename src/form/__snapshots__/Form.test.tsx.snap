// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form shallow render 1`] = `
<Fragment>
  <form
    onSubmit={[Function]}
  >
    <BaseInput
      id="first"
      key="first"
      label="count"
      type="number"
    />
    <DateInput
      id="datefield"
      key="datefield"
      label="DATE"
      type="date"
      value="2021-04-08T23:35:58.282Z"
    />
    <BaseInput
      id="second"
      key="second"
      label="name"
      type="text"
      value="hello"
    />
    <TextArea
      id="third"
      key="third"
      label="textarea"
      type="textarea"
      value="many text"
    />
    <CheckboxInput
      id="cb"
      key="cb"
      label="checkbox"
      type="checkbox"
      value={true}
    />
    <RadioButtons
      id="typeOfContact"
      items={
        Array [
          Object {
            "label": "mail",
            "value": "mail",
          },
          Object {
            "label": "phone",
            "value": "phone",
          },
        ]
      }
      key="typeOfContact"
      type="radio"
    />
    <div
      className="input-buttons"
    />
  </form>
</Fragment>
`;
